statemachine/statemachine.go

```diff
@@ -0,0 +1,316 @@
+		metafora.Debugf("task=%q in state %s", s.taskID, state.Code)
+		msg := s.exec(state)
+
+		// Enter State
+		// Apply Message
+		newstate, ok := apply(state, msg)
+		if !ok {
+			metafora.Warnf("task=%q Invalid state transition=%q returned by task. Old state=%q", s.taskID, msg.Code, state.Code)
+			msg = Message{Code: Error, Err: err}
+			if newstate, ok = apply(state, msg); !ok {
+				metafora.Errorf("task=%q Unable to transition to error state! Exiting with state=%q", s.taskID, state.Code)
+				return state.Code.Terminal()
+			}
+		}
+
+		// Save state
+		if err := s.ss.Store(s.taskID, newstate); err != nil {
+			metafora.Errorf("task=%q Unable to persist state=%q. Unscheduling.", s.taskID, newstate.Code)
+			return true
+		}
+
+		// Set next state and loop if non-terminal
+		state = newstate
+
+		// Exit and unschedule task on terminal state.
+		if state.Code.Terminal() {
+			return true
+		}
+
+		// Release messages indicate the task should exit but not unschedule.
+		if msg.Code == Release {
+			return false
+		}
+	}
+}
+
+// execute non-terminal states
+func (s *stateMachine) exec(state *State) Message {
+	switch state.Code {
+	case Runnable:
+		// Runnable passes control to the stateful handler
+		return run(s.h, s.taskID, s.cmds)
+	case Paused:
+		// Paused until a message arrives
+		return <-s.cmds
+	case Sleeping:
+		// Sleeping until the specified time (or a message)
+		if state.Until == nil {
+			metafora.Warnf("task=%q told to sleep without a time. Resuming.", s.taskID)
+			return Message{Code: Run}
+		}
+		dur := state.Until.Sub(time.Now())
+		metafora.Infof("task=%q sleeping for %s", s.taskID, dur)
+		timer := time.NewTimer(dur)
+		select {
+		case <-timer.C:
+			return Message{Code: Run}
+		case msg := <-s.cmds:
+			timer.Stop()
+			// Checkpoint is a special case that shouldn't affect sleep time, so
+			// maintain it across the state transition
+			if msg.Code == Checkpoint {
+				msg.Until = state.Until
+			}
+			return msg
+		}
+	case Fault:
+		// Special case where we potentially trim the current state to keep
+		// errors from growing without bound.
+		var msg Message
+		msg, state.Errors = s.errHandler(s.taskID, state.Errors)
+		return msg
+	default:
+		panic("invalid state: " + state.String())
+	}
+}
+
+func run(f StatefulHandler, tid string, cmd <-chan Message) (m Message) {
+	defer func() {
+		if r := recover(); r != nil {
+			metafora.Errorf("task=%q Run method panic()d! Applying Error message. Panic: %v", tid, r)
+			m = Message{Code: Error, Err: fmt.Errorf("panic: %v", r)}
+		}
+	}()
+	m = f(tid, cmd)
+	return m
+}
+
+// Stop sends a Release message to the state machine through the command chan.
+func (s *stateMachine) Stop() {
+	s.cmds <- Message{Code: Release}
+}
+
+// apply a message to cause a state transition. Returns false if the state
+// transition is invalid.
+func apply(cur *State, m Message) (*State, bool) {
+	//XXX Is a linear scan of all rules really the best option here?
+	for _, trans := range Rules {
```

Comment:
I was just going to make some comment about nested maps, (or an aliased type + helper method), but I don't think it would make anything clearer. :+1: